File: Tirggers/IoTHub_Trigger.cs
using IoTHubTrigger = Microsoft.Azure.WebJobs.EventHubTriggerAttribute;

using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Host;
using Microsoft.Azure.EventHubs;
using System.Text;
using System.Net.Http;
using Microsoft.Extensions.Logging;
using Microsoft.Azure.WebJobs.Extensions.DurableTask;
using System.Threading.Tasks;
using Microsoft.Function.Orchestrations.BlobJob;

namespace Microsoft.Function.Triggers
{
    public class IoTHub_Trigger
    {
        private static HttpClient client = new HttpClient();
        
        //IoTHub_Tirgger 는 Connection에 정의된 Endpoint의 IoT Hub 에서 이벤트가 발생하면 실행되며, 이 트리거 함수 자체가 오케스트레이션 클라이언트가 됨
        [FunctionName("IoTHub_Trigger")]
        public async Task Run(
            [IoTHubTrigger("messages/events", Connection = "IoTHubConnectionString")]EventData message, 
            [DurableClient]IDurableClient orchestratorClient,
            ILogger log)
        {

            var strMessage = Encoding.UTF8.GetString(message.Body.Array);

            log.LogInformation($"C# IoT Hub trigger function processed a message: {strMessage}");

            //Orchestrator 이름
            //var orchestratorName = "BlobJobOrchestrator";
            var orchestratorName = nameof(BlobJobOrchestrator); //Function 이름이 변경될수도 있으므로 nameof를 사용한다.

            //전달 할 파라메타
            var orchestratorInput = strMessage;

            var instanceId = await orchestratorClient.StartNewAsync(
                    orchestratorName,
                    orchestratorInput);

            //Orchestrator 인스턴스 ID
            log.LogInformation($"Started Orchestrator with ID = '{instanceId}'...");

        }
    }
}

File: python/Triggers/iothub_trigger.py
import logging
import json
import azure.functions as func
import azure.durable_functions as df


async def main(
    event: func.EventHubEvent,
    starter: str
) -> None:
    """
    IoTHub_Trigger - IoT Hub의 이벤트를 받아 Durable Function Orchestrator를 시작
    
    Args:
        event: IoT Hub에서 전달된 EventHub 이벤트
        starter: Durable Orchestration Client
    """
    # 메시지 파싱
    message_body = event.get_body().decode('utf-8')
    logging.info(f'Python IoT Hub trigger function processed a message: {message_body}')
    
    # Durable Orchestration Client 생성
    client = df.DurableOrchestrationClient(starter)
    
    # Orchestrator 시작
    instance_id = await client.start_new(
        orchestration_function_name='blob_job_orchestrator',
        client_input=message_body
    )
    
    logging.info(f"Started Orchestrator with ID = '{instance_id}'...")
